In [1]:
import pandas as pd
from neuronunit.optimization.optimization_management import public_opt
from sciunit.scores import RelativeDifferenceScore
from sciunit.utils import config_set, config_get
config_set('PREVALIDATE', False)
from neuronunit.allenapi.allen_data_driven import make_allen_hard_coded_limited, make_allen_hard_coded_complete
import warnings
from neuronunit.optimization.model_parameters import BPO_PARAMS
warnings.filterwarnings("ignore")
[allen_suite_623960880,allen_suite_623893177,_,_,_,df] = make_allen_hard_coded_complete()

/home/user/anaconda3/lib/python3.7/site-packages/cerberus/validator.py:1609: UserWarning: No validation schema is defined for the arguments of rule 'not_zero_obs_zscore'
  "'%s'" % method_name.split('_', 2)[-1]


# In this notebook
we fit Izhikevich and Adaptive Exponential Integrate and fire models to electrophysiology observations 
from the Allen Brain, using BluePyOpt and NeuronUnit.

* Immediately Below 
is a list of some of the observations we will use:

In [2]:
df

TimeConstantTest       name RestingPotentialTest  \
623960880          23.8 ms  623960880             -65.1 mV   
623893177          27.8 ms  623893177             -77.0 mV   
471819401          13.8 ms  471819401             -77.5 mV   
482493761          24.4 ms  482493761             -71.6 mV   

          InputResistanceTest RheobaseTest  
623960880       241.0 megaohm      70.0 pA  
623893177       136.0 megaohm     190.0 pA  
471819401       132.0 megaohm     190.0 pA  
482493761       132.0 megaohm      70.0 pA

Allen Specimen ID: 471819401
Four data fitting constraints:
* RheobaseTest	
* TimeConstantTest	
* RestingPotentialTest	
* InputResistanceTest	
# Model AdExp

In [3]:
_,_,_,a_cells = make_allen_hard_coded_limited()
opt,obs_preds,df = public_opt(
    a_cells['471819401'],
    BPO_PARAMS,
    '471819401',
    "ADEXP",
    100,
    300,
    "IBEA",
    score_type=RelativeDifferenceScore
    )


INFO:__main__:Run stopped because of stopping criteria: Max ngen
Run stopped because of stopping criteria: Max ngen


# Display the results of Optimization:
* We use the terms:
observations and predictions to relate the how closely the fitted model paramters match to the constraining ephys experimental data.
* observations: electrical feature values measured in data.
* predictions: outputs from fitted models that correspond to said features.

In [4]:
df

observations    predictions
RheobaseTest               190.0 pA      189.73 pA
TimeConstantTest            13.8 ms       13.85 ms
RestingPotentialTest       -77.5 mV      -77.48 mV
InputResistanceTest   132.0 megaohm  132.0 megaohm

# Solution Model Parameters:


In [5]:
opt.attrs_to_params()
params = pd.DataFrame([opt.params])
params

{'cm': 173.34,
 'v_spike': -22.17,
 'v_reset': -68.72,
 'v_rest': -77.48,
 'tau_m': 49.4,
 'a': 3.79,
 'b': 12.78,
 'delta_T': 4.31,
 'tau_w': 40.02,
 'v_thresh': -36.46,
 'spike_delta': 88.56}

Allen Specimen ID: 471819401
Four data fitting constraints:
* RheobaseTest	
* TimeConstantTest	
* RestingPotentialTest	
* InputResistanceTest	
# Model Izhikevich

In [6]:
opt,obs_preds,df = public_opt(
    a_cells['471819401'],
    BPO_PARAMS,
    '471819401',
    "IZHI",
    100,
    360,
    "IBEA",
    )

INFO:__main__:Run stopped because of stopping criteria: Max ngen
Run stopped because of stopping criteria: Max ngen


In [7]:
df

observations    predictions
RheobaseTest               190.0 pA       190.0 pA
TimeConstantTest            13.8 ms        6.61 ms
RestingPotentialTest       -77.5 mV      -73.09 mV
InputResistanceTest   132.0 megaohm  81.17 megaohm

# Solution Parameters:

In [8]:
opt.attrs_to_params()
params = pd.DataFrame([opt.params])
params

{'C': 157.71,
 'k': 0.9,
 'vr': -74.91,
 'vt': -45.84,
 'vPeak': 24.22,
 'a': 0.03,
 'b': 2.35,
 'c': -54.54,
 'd': 41.55,
 'celltype': 5}

# Just FI curve on AdExp model
Allen Specimin Id: 471819401

In [ ]:
opt,obs_preds,df = public_opt(
    a_cells['fi_curve'],
    BPO_PARAMS,
    'fi_curve',
    "ADEXP",
    100,
    300,
    "IBEA"
    )

In [ ]:
df


Allen Specimen ID: 623960880
Four data fitting constraints:
* RheobaseTest	
* TimeConstantTest	
* RestingPotentialTest	
* InputResistanceTest	
# Model Izhikevich

In [ ]:

opt,obs_preds,df = public_opt(
    allen_suite_623960880.tests,
    BPO_PARAMS,
    allen_suite_623960880.name,
    "IZHI",
    100,
    300,
    "IBEA"
    )

In [ ]:
df

# Solution parameters

In [ ]:
opt.attrs_to_params()
params = pd.DataFrame([opt.params])
params

In [ ]:

opt,obs_preds,df = public_opt(
    allen_suite_623893177.tests,
    BPO_PARAMS,
    allen_suite_623893177.name,
    "ADEXP",
    100,
    300,
    "IBEA"
    )

In [ ]:
df

# Solution parameters

In [ ]:
opt.attrs_to_params()
params = pd.DataFrame([opt.params])
params